rest/auth.go

```diff
@@ -35,29 +50,31 @@ type tokenResponse struct {
 }
 
 type TokenRequest struct {
-	ID         string `json:"id"`
-	TTL        int    `json:"ttl"`
-	Capability string `json:"capability"`
-	ClientID   string `json:"client_id"`
-	Timestamp  int64  `json:"timestamp"`
-	Nonce      string `json:"nonce"`
-	Mac        string `json:"mac"`
+	ID         string     `json:"id"`
+	TTL        int        `json:"ttl"`
+	Capability Capability `json:"capability"`
+	ClientID   string     `json:"client_id"`
+	Timestamp  int64      `json:"timestamp"`
+	Nonce      string     `json:"nonce"`
+	Mac        string     `json:"mac"`
 }
 
-func (t *TokenRequest) Sign(secret string) {
-	params := []string{
-		t.ID,
-		strconv.Itoa(t.TTL),
-		t.Capability,
-		t.ClientID,
-		strconv.Itoa(int(t.Timestamp)),
-		t.Nonce,
-	}
-	s := strings.Join(params, "\n") + "\n"
+var nl = []byte{'\n'}
 
+func (t *TokenRequest) Sign(secret string) {
 	mac := hmac.New(sha256.New, []byte(secret))
-	mac.Write([]byte(s))
-
+	mac.Write([]byte(t.ID))
+	mac.Write(nl)
+	mac.Write([]byte(strconv.Itoa(t.TTL)))
+	mac.Write(nl)
+	mac.Write([]byte(t.Capability.String()))
+	mac.Write(nl)
+	mac.Write([]byte(t.ClientID))
+	mac.Write(nl)
+	mac.Write([]byte(strconv.Itoa(int(t.Timestamp))))
+	mac.Write(nl)
+	mac.Write([]byte(t.Nonce))
+	mac.Write(nl)
```

Comment:
Hum... isn't replacing the array and the string a micro optimization? Also it's less readable IMO.